### DISCUSSION TITLE
Multiple DCUTR connections between two peers

### DISCUSSION THREAD

**Discussion by geotro** [libp2p/rust-libp2p#2881] [Q&A]
I'm a beginner libp2p user, specifically using the rust version.

Is it possible to start multiple DCUTR between two peers to achieve more than one direct connection behind NAT?

Thx

**Comment by thomaseizinger** [comment]
In principle, yes. You might need to open the connections sequentially though, not sure if it will work in parallel. @mxinden can likely answer this question in more detail.

Have you tried it already? Are you running in an error / some kind of problem?

**Reply by dariusc93** [reply]
if youre doing it with 2 instances of libp2p behind both nat that should work fine (ive done it with testing a few projects), although I never tried it with multiple connections to the same peer (assuming this is what youre implying) though more information on your use case might be helpful.

**Comment by mxinden** [comment]
@geotro would you mind expanding on why you want more than one connection between the same two peers?

**Comment by geotro** [comment]
@thomaseizinger Must admit I didn't try it yet, will do soon and revert.

@mxinden I want to do this in order to be sure I can utilize all the available bandwidth between two nodes

**Reply by mxinden** [reply]
What makes you think that a single TCP connection can not saturate your bandwidth? Also what is your use-case that you need to saturate your bandwidth? And lastly have you tested whether the bandwidth usage of a single connection is good enough?

**Reply by geotro** [reply]
The use case is transferring large amounts of data, as quickly as possible, between two peers.   Experience shows that a single TCP connection cannot usually saturate the available bandwidth over the internet, probably due to congestion control algorithms, on local networks it's a different story of course.

We did achieve pretty good single connection data rates with libp2p between two peers in different countries, but still far from using most of the bandwidth available to both peers.

The next step is to create as many parallel connections as necessary to achieve that.

**Reply by rkuhn** [reply]
This report is interesting in that it matches my experiments done in 2005 between research sites (TUM and CERN), where routing was across the GEANT network, so should arguably have been optimised for this use-case. The endpoints had GigE fiber connection more or less directly to the backbone.

One thing I also learnt back then is that TCP over such an ethernet/STM setup with a half dozen routers yields an undetected data corruption rate of the order of 1 per TB, which is “handled” by dropping the connection on the higher level. The lesson is that higher level segmentation should take this into account — it seems difficult to imagine that libp2p could grow the ability to recover from data corruption while keeping the same connection (it would be much easier when using UDP→QUIC).

**Reply by mxinden** [reply]
> yields an undetected data corruption rate of the order of 1 per TB

:open_mouth: is this still the case today? What was the root cause?

**Reply by mxinden** [reply]
> The use case is transferring large amounts of data, as quickly as possible, between two peers. Experience shows that a single TCP connection cannot usually saturate the available bandwidth over the internet, probably due to congestion control algorithms, on local networks it's a different story of course.
> 
> We did achieve pretty good single connection data rates with libp2p between two peers in different countries, but still far from using most of the bandwidth available to both peers.
> 
> The next step is to create as many parallel connections as necessary to achieve that.

Cool. I am curious what your experiments show. Would you mind posting the results here once you gathered the data? Likely useful for others as well. @geotro.

**Reply by rkuhn** [reply]
> 😮 is this still the case today? What was the root cause?

Well, ethernet frames have a bit error rate, TCP checksum is only 16bit, so I’d say something like this is to be expected.

**Comment by mxinden** [comment]
> Is it possible to start multiple DCUTR between two peers to achieve more than one direct connection behind NAT?

This is not possible today. `libp2p-dcutr` will not attempt a second hole punch in case a direct connection already exists. Forcing `libp2p-dcutr` to do so manually would probably be a small patch-set only. Though, by the contract of the protocol, this needs to be done by the listener side of the relayed connection.

**Reply by geotro** [reply]
@mxinden sure, would be happy to test and report back.  Could you point me in the right direction of what needs to be modified to allow this?

**Reply by mxinden** [reply]
You would want to add a method to the dcutr [`Behaviour`](https://github.com/libp2p/rust-libp2p/blob/master/protocols/dcutr/src/behaviour.rs) which does the same as when an incoming relayed connection is handled:

https://github.com/libp2p/rust-libp2p/blob/5906140d38b4298a97a6b3cec9cb9a41962463c0/protocols/dcutr/src/behaviour.rs#L108-L133

This method you can then call from the `Swarm::behaviour_mut()` to trigger the next DCUtR process.

**Comment by geotro** [comment]
Thank you.  Hoping to be able to test this soon. Will report back with the results.
